docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+On the other hand in the current system, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Very often, pods don’t use all the resources that they request, which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a system with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in another 4 pods of this type, but Kubernetes will reject new pods.
+
+In this design document, we propose **implementing two tiers of pods** - one in which we guarantee resources (top-tier pods) and the second wherein we provide resources if they’re available and at a slightly lower quality (best-effort pods). Adding a best-effort tier of pods will increase node’s utilization in a Kubernetes cluster (in Borg, supporting best-effort pods increased utilization by  about 20%). We also implement **a policy of automatically classifying pods that don’t provide limits as best-effort**. This provides much better resource guarantees to pods with limits specified.
+
+## QoS Functionality
+
+We will introduce two tiers of pods: top-tier pods and best-effort pods. We will not support multiple levels or priorities.
+
+Best-effort pods will be invisible to top-tier pods:
+- **Existence of best-effort pods should not prevent top tier pods from launching**. When deciding whether to launch a top-tier pod, we only look at resource limits of other top-tier pod. Suppose we have a machine with 4GB of memory. Top-tier pod P1 has a 2GB limit, and 4 best-effort pods have a 500MB limit each. Suppose the user launches a new top-tier pod P2 with a 1GB limit. The Kubelet will evict best-effort pods, in unspecified order, to make room for P2. P2 will then be admitted and allocated the requested resources.
+- **Allocation of excess compressible resources (e.g. CPU) to top-tier pods will be the same as if best-effort pods did not exist**. For example if top-tier pod P1 asks for minimum 20% CPU, and we also have best-effort pods B1 and B2 running. Suppose P1 needs more CPU (because of a huge number of requests coming in), then P1 will starve B1, B2 of CPU and use the whole 100% of the CPU (assuming there are no other top tier pods running).
+- **Allocation of non-compressible resources (e.g. memory) to top-tier pods will be the same as if best-effort pods did not exist**. For example consider a system with 4GB of memory, where top-tier pod P1 is using 500MB of memory (with a 1GB limit) and best-effort pods are using the other 3.5GB of memory. The best-effort pods will not prevent P1 from accessing more memory (assuming that P1 remains within its memory limits). So if P1 asks for 100MB more memory, P1 will get the memory it needs, and one of the other pods will be OOM killed. 
+
+Resource guarantees for best-effort pods:
+- Compressible resources
+  - For now, we are only implementing CPU limits.

```

Comment:
Right now, our CPU limits converts to ```--cpu-shares``` which only have meaning when multiple containers are competing for processor time.

I am interested in also allowing cluster operators to enable ```--cpu-quota``` to provide a ceiling to the maximum amount of CPU a given container/pod can consume before its actively throttled.

The rationale is that some cluster operators want to provide a stable level of cpu performance when there is 1 versus 100 containers on the same node to their end user.

This is worthy of its own issue discussion which I will open, but I think it also relates to QoS tiers in general so noting it here so it can be tracked as a broader community concern.